LimitRange example improvements, detailed README

derekwaynecarr opened the pull request:
Fixes #5149 
1. Adds a detailed README for when/why you would use a `LimitRange`.
2. Moves example files to yaml, ensured new files were caught in example_test
3. Has the right analytics water-mark

derekwaynecarr committed:
LimitRange example improvements, detailed README

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
-Please refer to this [doc](https://github.com/GoogleCloudPlatform/kubernetes/blob/620af168920b773ade28e27211ad684903a1db21/docs/design/admission_control_limit_range.md#kubectl).
+Limit Range
+========================================
+By default, a pod will run with unbounded CPU and memory limits.  This means that any pod in the
```
nit: "pods run" rather?

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+1. Each node in the cluster has 2GB of memory.  The cluster operator may want to reject any pod that requests
+more than that amount prior to admitting the pod in the cluster since that pod could never be scheduled by the
+system.  As a result, an operator has a desire to set the maximum amount of resource a single pod could consume
+on a system in order to increase the odds that a pod could fit on a node in the cluster.
```
nit: This is not technically accurate, right?  Just adding the limit does not affect schedulability?  It just improves error reporting.  Worth clarifying?

derekwaynecarr replied:
Agree on the confusion there, will clarify this phrasing.

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+more than that amount prior to admitting the pod in the cluster since that pod could never be scheduled by the
+system.  As a result, an operator has a desire to set the maximum amount of resource a single pod could consume
+on a system in order to increase the odds that a pod could fit on a node in the cluster.
+2. The cluster is shared by multiple communities in an organization to run both production and development
```
nit: Perhaps simplifiy this for better readability?  e.g. something like "A cluster is shared by two communities in an organization which run production and development workloads respectively."

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+separated from production community workloads by a Kubernetes namespace.  The development namespace accepts pods
+that consume no more than 512MB of memory, but the production namespace accepts pods that consume up to 8GB of
+memory.  To support each mode, the cluster operator applies separate limits to each namespace.
+3. Users may create a Pod which consume resources just below the capacity of a machine.  The left over space
```
typo: "which consumeS"

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+the cluster operator may want to set limits that a pod must consume at least 20% of the memory and cpu of their
+average node size in order to provide for more uniform scheduling and to limit waste.
+
+This example assumes you have a functional Kubernetes setup.
```
nit: "... assumes THAT you..."

quinton-hoole commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+
+1. A container in a pod may consume between 6Mi and 1Gi of memory.  If the container does not
+specify an explicit resource limit, each container in a pod will get 100Mi of memory.
+2. A container in a pod may consume between 250m and 2 cores of cpu.  If the container does
```
nit: Might be worth clarifying the meaning of the 'm' suffix.  I assume that it's milli-cpu.

derekwaynecarr replied:
It is milli cpu.  I am adding a link to the resources model in the introduction.

quinton-hoole commented:
LGTM barring very minor stylistic nits.  @derekwaynecarr I assume that you've run this manually to validate?

derekwaynecarr commented:
@quinton-hoole - will make style updates, and yes, I ran this example as part of collecting the output based on HEAD from this morning.  will ping when updates are made.

quinton-hoole commented:
Ping @derekwaynecarr ?

derekwaynecarr commented:
Will fix tomorrow.  Had other pressing issues to handle today.  Thanks for the reminder. 

Sent from my iPhone

> On Jun 8, 2015, at 6:59 PM, Quinton Hoole <notifications@github.com> wrote:
> 
> Ping @derekwaynecarr ?
> 
> —
> Reply to this email directly or view it on GitHub.
>

quinton-hoole commented:
Ping? @derekwaynecarr 
No desperate rush from my side, just want to make sure that this PR goes in eventually.

derekwaynecarr commented:
nits addressed

derekwaynecarr commented:
@quinton-hoole - sent this back your way since comments were addressed and I don't want to self tag a PR with lgtm.

roberthbailey commented:
@ihmccreery can you take a look through the README.md file and verify that it works on GCE/GKE?

ihmccreery commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+the cluster operator may want to set limits that a pod must consume at least 20% of the memory and cpu of their
+average node size in order to provide for more uniform scheduling and to limit waste.
+
+This example assumes that you have a functional Kubernetes setup.
```
Please link to Getting Started as other examples do.

derekwaynecarr replied:
adding a step 0 to link to that.

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+
+Let's describe the limits that we have imposed in our namespace.
+
+```shell
````
Please add the command, (I believe it's `$ k describe limits limits --namespace=limit-example`, but I could be wrong).

derekwaynecarr replied:
done

bgrant0607 replied:
It would be less confusing if the object were not called "limits". Perhaps "mylimits".

ihmccreery commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+Name:   limits
+Type    Resource  Min Max Default
+----    --------  --- --- ---
+Pod   memory    6Mi 1Gi -
```
Please format these properly.

ihmccreery commented on examples/limitrange/README.md:
```diff
@@ -1,4 +1,147 @@
+2. A container in a pod may consume between 250m and 2 cores of cpu.  If the container does
+not specify an explicit resource limit, each container in a pod will get 250m of cpu.
+3. The total memory usage of a pod across all of its containers must fall between 6Mi and 1Gi.
+4. The total cpu usage of a pod across all of its containers must fall between 250m and 2 cores.
```
It would be clearer if these statements were in the same order as the line-items above; i.e. pod/mem, pod/cpu, container/mem, container/cpu.

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+Let's create a new namespace called limit-example:
+
+```shell
+$ cluster/kubectl.sh create -f namespace.yaml
````
This should be run from the kubernetes root, right?  So it should be `-f examples/limitrange/namespace.yaml`.

derekwaynecarr replied:
I will add a blurb to change to the limit range example directory.  Also will remove cluster/kubectl.sh in favor of just kubectl

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+Let's create a simple limit in our namespace.
+
+```shell
+$ cluster/kubectl.sh create -f limits.yaml --namespace=limit-example
````
Same as above: `-f examples/limitrange/limits.yaml`.

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+```
+
+Note that this pod specifies explicit resource limits so it did not pick up the namespace default values.
+
````
It would be good to have a **Step 4: Cleanup** as well.

ihmccreery commented:
@derekwaynecarr: Overall, excellent write-up.  Thanks so much for the clear writing; it's really awesome to see a tutorial written with such good explanations.

I added some inline comments for you; otherwise, this looks good to go.  I ran through it on GKE; I still need to test it for GCE, but I'm having some trouble turning up a cluster.  Once I get that sorted out, I can finish checking a run through on GCE and give you a LGTM.

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+
+```shell
+$ cluster/kubectl.sh create -f namespace.yaml
+$ cluster/kubectl.sh get namespaces
````
Please include the expected output from the prior command before this.

ihmccreery commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,147 @@
+
+```shell
+$ cluster/kubectl.sh create -f limits.yaml --namespace=limit-example
+```
````
Also same as above: please include expected output.

ihmccreery commented:
Finished run-through with GCE.  LGTM, once comments are addressed.

quinton-hoole commented:
LGTM

quinton-hoole commented:
@bgrant0607 can you give a second LGTM and ok-to-merge?

derekwaynecarr commented:
I will make the updates requested tomorrow. Thanks

Sent from my iPhone

> On Jun 25, 2015, at 4:30 PM, Isaac Hollander McCreery <notifications@github.com> wrote:
> 
> @derekwaynecarr: Overall, excellent write-up. Thanks so much for the clear writing; it's really awesome to see a tutorial written with such good explanations.
> 
> I added some inline comments for you; otherwise, this looks good to go. I ran through it on GKE; I still need to test it for GCE, but I'm having some trouble turning up a cluster. Once I get that sorted out, I can finish checking a run through on GCE and give you a LGTM.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on examples/limitrange/valid-pod.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1beta3
```
Why is this not v1?

bgrant0607 commented on examples/limitrange/valid-pod.yaml:
```diff
@@ -0,0 +1,13 @@
+  name: valid-pod
+spec:
+  containers:
+  - capabilities: {}
```
Remove default/empty fields.

bgrant0607 commented on examples/limitrange/valid-pod.yaml:
```diff
@@ -0,0 +1,13 @@
+  containers:
+  - capabilities: {}
+    image: nginx
+    name: nginx
```
Put name before image.

bgrant0607 commented on examples/limitrange/namespace.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1beta3
```
v1

bgrant0607 commented:
Please update everything to the v1 API.

derekwaynecarr commented:
Addressed all feedback on text.

1. moved to just ```kubectl``` instead of ```cluster/kubectl.sh```
2. ran through scenario on GCE provider to ensure all is good
3. updated examples to v1
4. removed empty values from pod definitions

derekwaynecarr commented:
@bgrant0607 - updated example to use "mylimits" instead of "limits"

bgrant0607 commented on examples/limitrange/README.md:
````diff
@@ -1,4 +1,166 @@
+
+```shell
+$ kubectl create -f limits.yaml --namespace=limit-example
+limitranges/mylimits
````
Sorry. You use "limit-example" as the namespace in most places. Shouldn't this be "limit-example" instead of "mylimits", or change all instances of "limit-exmaple" to "mylimits".

Also, namespaces are namespaced, so I assume the --namespace on the command line shouldn't do anything.

bgrant0607 replied:
Nevermind.....

bgrant0607 commented:
LGTM

bgrant0607 commented:
ok to merge

previous LGTM was here: https://github.com/GoogleCloudPlatform/kubernetes/pull/9264#issuecomment-115398302

bgrant0607 commented:
Shippable failed with:
```
the path "examples/limitrange/valid-pod.json" does not exist
!!! Error in ./hack/test-cmd.sh:230
```

bgrant0607 commented:
Trying to get this in today, if possible. Rerunning shippable.

bgrant0607 commented:
@k8s-bot test this please

bgrant0607 commented:
LGTM and ok to merge, assuming tests pass. We have a new link checker, as of a few minutes ago, so I want to make sure we run through that.

bgrant0607 commented:
Indeed:
```
./hack/verify-gendocs.sh
diffing /root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/docs/ against freshly generated docs
Some md files are missing ga-beacon analytics link:
/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/docs/devel/cherry-picks.md
/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/docs/ is out of date. Please run hack/run-gendocs.sh
```

bgrant0607 commented:
Not from this PR. Will find out what broke.

bgrant0607 commented:
Looks like PR #10709

bgrant0607 commented:
Fixed by #11073. Restarting shippable again.

rjnagal commented:
shippable is still complaining about gendocs.

bgrant0607 commented:
No longer complaining about gendocs, but may have hit a flaky test.

rjnagal commented:
merging for the cut.
